VirtualReality/Widgets/qMRMLVirtualRealityTransformWidget.cxx

```diff
@@ -0,0 +1,364 @@
+  limitations under the License.
+
+==============================================================================*/
+
+// Slicer includes
+#include <qSlicerAbstractCoreModule.h>
+#include <qSlicerApplication.h>
+#include <qSlicerModuleManager.h>
+
+// qMRML includes
+#include "qMRMLVirtualRealityTransformWidget.h"
+#include "ui_qMRMLVirtualRealityTransformWidget.h"
+
+// MRML includes
+#include <vtkMRMLLinearTransformNode.h>
+
+// OpenVR includes
+#include <openvr.h>
+
+// SlicerVR includes
+#include <qMRMLVirtualRealityView.h>
+#include <vtkMRMLVirtualRealityViewNode.h>
+
+// Qt includes
+#include <QDebug>
+#include <QPushButton>
+
+namespace
+{
+  vr::ETrackingResult StringToPoseStatus(const char* cpose)
+  {
+    if (cpose == NULL)
+    {
+      return vr::TrackingResult_Uninitialized;
+    }
+    std::string pose(cpose);
+
+    if (pose.compare("CalibratingInProgress") == 0)
+    {
+      return vr::TrackingResult_Calibrating_InProgress;
+    }
+    else if (pose.compare("CalibratingOutOfRange") == 0)
+    {
+      return vr::TrackingResult_Calibrating_OutOfRange;
+    }
+    else if (pose.compare("RunningOk") == 0)
+    {
+      return vr::TrackingResult_Running_OK;
+    }
+    else if (pose.compare("RunningOutOfRange") == 0)
+    {
+      return vr::TrackingResult_Running_OutOfRange;
+    }
+    else
+    {
+      return vr::TrackingResult_Uninitialized;
+    }
+  }
+}
+//-----------------------------------------------------------------------------
+/// \ingroup Slicer_QtModules_Markups
+class qMRMLVirtualRealityTransformWidgetPrivate
+  : public Ui_qMRMLVirtualRealityTransformWidget
+{
+  Q_DECLARE_PUBLIC(qMRMLVirtualRealityTransformWidget);
+protected:
+  qMRMLVirtualRealityTransformWidget* const q_ptr;
+public:
+  qMRMLVirtualRealityTransformWidgetPrivate(qMRMLVirtualRealityTransformWidget& object);
+  void init();
+
+  vtkWeakPointer<vtkMRMLVirtualRealityViewNode> VRViewNode;
+  vtkWeakPointer<vtkMRMLLinearTransformNode>    TransformNode;
+  vr::ETrackedDeviceClass                       TransformType;
+  vr::ETrackingResult                           PreviousStatus;
+};
+
+//-----------------------------------------------------------------------------
+// qMRMLVirtualRealityTransformWidgetPrivate methods
+
+//-----------------------------------------------------------------------------
+qMRMLVirtualRealityTransformWidgetPrivate::qMRMLVirtualRealityTransformWidgetPrivate(qMRMLVirtualRealityTransformWidget& object)
+  : q_ptr(&object)
+  , VRViewNode(nullptr)
+  , TransformNode(nullptr)
+  , TransformType(vr::TrackedDeviceClass_Invalid)
+  , PreviousStatus(vr::TrackingResult_Uninitialized)
+{
+
+}
+
+//-----------------------------------------------------------------------------
+void qMRMLVirtualRealityTransformWidgetPrivate::init()
+{
+  Q_Q(qMRMLVirtualRealityTransformWidget);
+  this->setupUi(q);
+
+  QObject::connect(this->pushButton_Transform, &QPushButton::clicked, q, &qMRMLVirtualRealityTransformWidget::onButtonClicked);
```

Comment:
Just wondering: is this some kind of new way of doing connections in Qt, or just a different way that existed before? (Usually we use SIGNAL(clicked()) and SLOT(onButtonClicked) )

Reply:
Yes, this it the new C++11 way of doing Qt connections. You can also embed lambdas directly into the SLOT side.

See [https://wiki.qt.io/New_Signal_Slot_Syntax](https://wiki.qt.io/New_Signal_Slot_Syntax)